[AMDGPU][LIT] Added a MIR LIT showing the SGPR spills

vg0204 opened the pull request:
This MIR test case is added to seek the consumption of VGPR lanes being used for SGPR spills during si-lower-sgpr-spills pass of AMDGPU pass pipeline. Basically, in this pass, stack slots are mapped to available VGPR lanes for spilling purpose, thus ending the need for stack slots.

In current scenario, each new SGPR spill goes into new VGPR lanes as, being mapped from its distinct stack slots assigned during SGPR allocation pass. It can be clearly seen in the added test case.

vg0204 committed:
[AMDGPU][LIT] Added a MIR LIT showing the SGPR spills

This MIR test case is added to seek the consumption of VGPR lanes
being used for SGPR spills during si-lower-sgpr-spills pass of
AMDGPU pass pipeline. Basically, in this pass, stack slots are
mapped to available VGPR lanes for spilling purpose, thus ending
the need for stack slots.

In current scenarion, each new SGPR spill goes into new VGPR lanes
as, being mapped from its distinct stack slots assigned during SGPR
allocation pass. It can be clearly seen in the added test case.

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -0,0 +1,231 @@
+# In the remaining 2 test cases mentioned in just above 2 lines, the stack slot indices is shared among the spill stack objects of
+# unequal size, with spill slot having the size of the largest of the stack objects sharing the common stack indices.
+
+--- |
```
You shouldn't need the IR section. The only thing the IR is providing you here is the required global callee for func, which you can substitute with a 0 in the MIR assuming the callee is insignificant (which it should be here)

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -0,0 +1,231 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -verify-machineinstrs -stress-regalloc=3 -start-before=greedy -stop-after=si-lower-sgpr-spills -o - %s | FileCheck -check-prefix=SGPR_SPILLED %s
```
This test would be more reliable if it did not depend on register allocation behavior. Can you only run the post-RA passes?

vg0204 replied:
I did not understand much what you asked me to do, Can you elaborate more?

cdevadas replied:
Change the RUN line to use --start-before=stack-slot-coloring --stop-after=si-lower-sgpr-spills. The original test you have written should be modified first. To use this new RUN line, the test should be the output of sgpr-regalloc pipeline just before invoking the newly introduced stack slot coloring.

vg0204 replied:
Now, understood, thanks for the clarification!

vg0204 replied:
As I was trying to generate post optimized SGPR spills by adding " --start-before=stack-slot-coloring --stop-after=si-lower-sgpr-spills", I found that if stackSlotColoring(SSC) is introduced before SILowerSGPRSpills, it really depends on regAlloc. As StackSlotColoring pass uses LiveStack(LS) analysis results {which is currently computed at RegAlloc phase, as LS itself do not do anything}. So if we invoke SSC directly on 'input.mir' without LS results available, we acutually won't see any stack slots optimization (in turn no sgpr spill optimization).

vg0204 replied:
So, I guess this test file need to consider to start from RA rather than post RA else, SCC really won't work thereafter(with optimized SGPR spill code) because of absence of LS , which really gets calculated at RA.

vg0204 replied:
The pipeline is stopped way before VGPR allocation here, as it happens much after SILowerSGPR spills. The -start-before=greedy, here it refers to SGPR greedy allocation, followed by virtRegRewriter & SILOwerSGPRSpill. I feel the change needed would eith to use -sgpr-regalloc option OR use  -start-before=greedy,0 to depict using SGPR allocation.

vg0204 replied:
Just like to add, VGPR allocation is happening after SGPR allocation already happened.

vg0204 replied:
@cdevadas, can you add your thoughts to it!

cdevadas replied:
I think it is reasonable to start the test from sgpr-regalloc and keep it until sgpr spill lowering to capture how the newly inserted stack slot coloring helps optimize vgpr lanes. I don't see any problem with the test.

vg0204 committed:
Modified the LIT test to eliminate dependency on RA behaviour.

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -8,44 +8,53 @@
+alignment:       1
 tracksRegLiveness: true
+tracksDebugUserValues: true
+registers:
+  - { id: 0, class: sreg_32_xm0_xexec }
+  - { id: 1, class: sreg_64_xexec }
+  - { id: 2, class: vreg_64 }
+  - { id: 3, class: vgpr_32, preferred-register: '$vgpr0' }
+  - { id: 4, class: sreg_64 }
+  - { id: 5, class: sreg_32_xm0_xexec }
+  - { id: 6, class: sreg_64_xexec }
+  - { id: 7, class: sreg_64 }
+  - { id: 8, class: sreg_64 }
+  - { id: 9, class: sreg_64 }
```
Remove the registers section

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -8,44 +8,53 @@
-
 ---
 name:            stack-slot-share-equal-sized-spills
+alignment:       1
```
Can remove this

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -8,44 +8,53 @@
+  - { id: 8, class: sreg_64 }
+  - { id: 9, class: sreg_64 }
 frameInfo:
+  maxAlignment:    4
```
Remove this

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -8,44 +8,53 @@
-  scratchRSrcReg: $sgpr0_sgpr1_sgpr2_sgpr3
-  frameOffsetReg: $sgpr32
-  stackPtrOffsetReg: $sgpr32
+  maxKernArgAlign: 1
```
Remove this

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -8,44 +8,53 @@
+  scratchRSrcReg:  '$sgpr0_sgpr1_sgpr2_sgpr3'
+  frameOffsetReg:  '$sgpr32'
+  stackPtrOffsetReg: '$sgpr32'
+  argumentInfo:
+    privateSegmentBuffer: { reg: '$sgpr0_sgpr1_sgpr2_sgpr3' }
+    dispatchPtr:     { reg: '$sgpr4_sgpr5' }
+    queuePtr:        { reg: '$sgpr6_sgpr7' }
+    dispatchID:      { reg: '$sgpr10_sgpr11' }
+    workGroupIDX:    { reg: '$sgpr12' }
+    workGroupIDY:    { reg: '$sgpr13' }
+    workGroupIDZ:    { reg: '$sgpr14' }
+    LDSKernelId:     { reg: '$sgpr15' }
+    implicitArgPtr:  { reg: '$sgpr8_sgpr9' }
+    workItemIDX:     { reg: '$vgpr31', mask: 1023 }
+    workItemIDY:     { reg: '$vgpr31', mask: 1047552 }
+    workItemIDZ:     { reg: '$vgpr31', mask: 1072693248 }
```
Remove this

arsenm commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -77,41 +86,81 @@ body:             |
+alignment:       1
 tracksRegLiveness: true
+tracksDebugUserValues: true
+registers:
```
Same in all these other functions, can remove most of this

vg0204 committed:
Removed unnecessary machineInfo from MIR LIT test.

vg0204 committed:
Changed MIR test to add dependency from RA

This test case is made to test post optimization SGPR spills which depends on StackSlotColoring pass which utilizes LiveStack analysis results. As for current scenarion LS analysis is really carried out at RA phase, making it necessary to be invoked in order for LS computation.

cdevadas commented:
Wait for Matt if he has any concerns.

vg0204 committed:
Added flag to indicate test invocation before SGPR allocation.

cdevadas commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -0,0 +1,206 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx908 -verify-machineinstrs -stress-regalloc=3 -start-before=greedy,0 -stop-after=si-lower-sgpr-spills -o - %s | FileCheck -check-prefix=SGPR_SPILLED %s
+
+# INFO : The pass pipeline is started from before the first iteration of register allocation pass that represents SGPR allocation in AMDGPU pipleine.
```
```suggestion
# INFO : The test starts from the sgpr-regalloc pipeline.
```

cdevadas commented on llvm/test/CodeGen/AMDGPU/si-lower-sgpr-spills-vgpr-lanes-usage.mir:
```diff
@@ -0,0 +1,206 @@
+# In the remaining 2 test cases mentioned in just above 2 lines, the stack slot indices is shared among the spill stack objects of
+# unequal size, with spill slot having the size of the largest of the stack objects sharing the common stack indices.
+
+
```
Extra line?

vg0204 committed:
Addressed changes in comments.
